src/coreclr/src/jit/lower.cpp

```diff
@@ -6365,3 +6314,189 @@ void Lowering::ContainCheckBitCast(GenTree* node)
+            // GT_NULLCHECK is a non-value node and would therefore require an internal register
+            // to use as the target. That is non-optimal because it will be modeled as conflicting
+            // with the source register(s).
+            // So, to summarize:
+            // - On ARM64, always use GT_NULLCHECK for a dead indirection.
+            // - On ARM, always use GT_IND.
+            // - On XARCH, use GT_IND if we have a contained address, and GT_NULLCHECK otherwise.
+            // In all cases, change the type to TYP_INT.
+            //
+            ind->gtType = TYP_INT;
+#ifdef TARGET_ARM64
+            bool useNullCheck = true;
+#elif TARGET_ARM
+            bool useNullCheck = false;
+#else  // TARGET_XARCH
+            bool useNullCheck = !ind->Addr()->isContained();
+#endif // !TARGET_XARCH
+
+            if (useNullCheck && ind->OperIs(GT_IND))
+            {
+                ind->ChangeOper(GT_NULLCHECK);
+                ind->ClearUnusedValue();
+            }
+            else if (!useNullCheck && ind->OperIs(GT_NULLCHECK))
+            {
+                ind->ChangeOper(GT_IND);
+                ind->SetUnusedValue();
+            }
+        }
+    }
+    else
+    {
+        // if `ADDR` node under `IND(struct(ADDR))` under `STORE_OBJ`
+        // is a complex one it could benefit from a not contained `LEA`.
+        TryCreateAddrMode(ind->Addr(), false);
+    }
+}
+
+//------------------------------------------------------------------------
+// LowerBlockStoreCommon: a common logic to lower STORE_OBJ/BLK/DYN_BLK.
+//
+// Arguments:
+//    blkNode - the store blk/obj node we are lowering.
+//
+void Lowering::LowerBlockStoreCommon(GenTreeBlk* blkNode)
+{
+    assert(blkNode->OperIs(GT_STORE_BLK, GT_STORE_DYN_BLK, GT_STORE_OBJ));
+    if (TryTransformStoreObjAsStoreInd(blkNode))
+    {
+        return;
+    }
+
+    LowerBlockStore(blkNode);
+}
+
+//------------------------------------------------------------------------
+// TryTransformStoreObjAsStoreInd: try to replace STORE_OBJ/BLK as STOREIND.
+//
+// Arguments:
+//    blkNode - the store node.
+//
+// Return value:
+//    true if the replacement was made, false otherwise.
+//
+// Notes:
+//    TODO: that method should do the transformation when possible
+//    and STOREIND should always generate better or the same code as
+//    STORE_OBJ/BLK for the same copy.
+//
+bool Lowering::TryTransformStoreObjAsStoreInd(GenTreeBlk* blkNode)
+{
+    assert(blkNode->OperIs(GT_STORE_BLK, GT_STORE_DYN_BLK, GT_STORE_OBJ));
+    if (blkNode->OperIs(GT_STORE_DYN_BLK))
+    {
+        return false;
+    }
+
+    ClassLayout* layout = blkNode->GetLayout();
+    if (layout == nullptr)
+    {
+        return false;
+    }
+
+    var_types regType = layout->GetRegisterType();
+    if (regType == TYP_UNDEF)
+    {
+        return false;
+    }
+    if (varTypeIsSIMD(regType))
+    {
+        // TODO: support STORE_IND SIMD16(SIMD16, CNT_INT 0).
```

Comment:
This and the `TODO` below should probably be `TODO-CQ`

Reply:
Done